### ISSUE TITLE
Wayland: sway master triggers protocol error message on exit

### ISSUE THREAD

**Issue by nohsheef** [fltk/fltk#750]
**Describe the bug**
On sway, anytime I close an FLTK window, I see this:

```
zxdg_toplevel_decoration_v1@32: error 2: xdg_toplevel destroyed before xdg_toplevel_decoration
Fatal error while communicating with the Wayland server: Protocol error
```

Pressing ESC triggers a controlled exit and, as expected, does not report an error.

Is sway's window close operation perhaps too disruptive and interpreted as a fatal event? If so, is that the default we want, or would an FLTK app with data to lose not already hook the event and prevent uncontrolled exiting?

I guess this is less bug and more RFC if we want to change the behavior.

1. Run your FLTK app
2. Close the window via sway's hotkey

**Expected behavior**
More graceful exit upon sway's regular window close operation.

**FLTK Version**
 - fltk-rs git master

**FLTK Configure / Build Options**
 - ./configure --disable-x11 --enable-wayland ...

**Operating System / Platform:**
 - Arch Linux

**Linux/Unix Runtime, if applicable:**
 - Wayland
 - sway

**Comment by Albrecht-S** [comment]
@nohsheef Thank you for the report. This sounds like something that should (and could hopefully) be fixed. All error messages cause by legit user handling should be avoided.

One question for better understanding (because I don't use sway myself): what is "sway's hotkey" to close the window?

I'm confident that @ManoloFLTK will be able to reproduce the issue because I believe that he has a sway test environment available.

**Comment by Albrecht-S** [comment]
Note: I classified this as a bug with prio 3 (moderate) for now.

**Comment by nohsheef** [comment]
I don't know if it should be classified as a real bug, but it's the regular close/kill command, which is by default bound to `$mod+Shift+q`.

From https://github.com/swaywm/sway/blob/master/sway/sway.5.scd:

> _kill_
> Kills (closes) the currently focused container and all of its children.

If I kill/close a Qt5, Qt6, or GTK3 window, I don't see an error mesage. In fact, when I kill qmmp, I can see it reporting terminating sync calls and other destructors/finalizers:

> Visual::~Visual()
> void QmmpUiSettings::sync()
> PlayListManager: saving playlists...
> void QmmpSettings::sync()

Therefore I assume sway's kill command triggers an event that can be subscribed to by applications that want to trigger a controlled exit.

**Comment by nohsheef** [comment]
I don't know if it should be classified as a real bug, but it's the regular close/kill command which is by default bound to `$mod+Shift+q`.

From https://github.com/swaywm/sway/blob/master/sway/sway.5.scd:

> _kill_
> Kills (closes) the currently focused container and all of its children.

If I kill/close a Qt5, Qt6, or GTK3 window, I don't see an error mesage. In fact, when I kill qmmp, I can see it report terminating sync calls and other destructors/finalizers:

> Visual::~Visual()
> void QmmpUiSettings::sync()
> PlayListManager: saving playlists...
> void QmmpSettings::sync()

Therefore I assume sway's kill command triggers an event that can be subscribed to by applications that want to trigger a controlled exit.

**Comment by Albrecht-S** [comment]
If I understand correctly, then there's a similar global shortcut on macOS: `command + q` which is handled by FLTK in a controlled way. I imagine this could be done with sway as well, but I'm only guessing.

**Comment by Albrecht-S** [comment]
As for classifying as a bug: such unwanted, annoying warnings/error messages should IMHO be avoided (suppressed) if possible, hence I'd say, from a user's perspective this is a bug. We can always re-classify this issue though if it turns out that we can't fix it on our side.

**Comment by ManoloFLTK** [comment]
I've tried to kill FLTK apps under the Sway compositor and see no error occurring and the app closed. My configuration
FreeBSD 13.2-RELEASE
FLTK current from the git repo

I'm not sure what the `$mod` key is on my Apple keyboard. I've therefore changed in file
`$HOME/.config/sway/config`
line
`bindsym $mod+Shift+q kill`
into
`bindsym Ctrl+Shift+q kill`
or
`bindsym Alt+Shift+q kill`

In both cases, the changed kill key does kill the FLTK app without error message.

At this point, I don't see how to reproduce what you report.

**Comment by nohsheef** [comment]
One difference is that I use the unreleased scene graph branch (https://github.com/swaywm/sway/pull/6844#commits-pushed-90c1b2c). Will test with 1.8.1

**Comment by nohsheef** [comment]
@Nefsen402 this happens with https://github.com/swaywm/sway/pull/6844/commits/b1a3960e095850d965b1c5bb75d17eec3c1c3e3d, but not 1.8.1. Any idea?

**Comment by Nefsen402** [comment]
> @Nefsen402 this happens with [swaywm/sway@b1a3960](https://github.com/swaywm/sway/commit/b1a3960e095850d965b1c5bb75d17eec3c1c3e3d), but not 1.8.1. Any idea?

This has nothing to do with the scene graph PR. Other work ends up landing on wlroots including many changes to how dxg shells are handled. Not everything revolves around that PR. I'm sure you'll find that the same bug happens on sway master (which is different from the last release).

**Comment by nohsheef** [comment]
@Nefsen402 you're right, and I apologize for implying that message. I considered the changes in the branch first before I considered unpatched sway master. Thank you for clarifying the branch has no changes related to it.

**Comment by nohsheef** [comment]
I'll test with plain sway master next before I file a ticket.

**Comment by Nefsen402** [comment]
It looks like this app is missing a call to [`zxdg_toplevel_decoration_v1::destroy`](https://wayland.app/protocols/xdg-decoration-unstable-v1#zxdg_toplevel_decoration_v1:request:destroy). This particular problem here started being reported with this wlroots PR: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/4051. This is an app bug BTW.

**Comment by nohsheef** [comment]
@Nefsen402 for the record, you were absolutely right. https://github.com/swaywm/sway/commit/91079079690c8832b22960403f0e60c63822d17a with wlroots [4de9c9c](https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/4de9c9c47f18cafb10b0a8110550e16b61c3093c) provokes the same error.

Sorry for jumping the gun without trying sway master first, and thanks a lot for taking time to investigate!

**Comment by ManoloFLTK** [comment]
@nohsheef Could you, please, try to apply the 1-line patch below to the FLTK source code, rebuild libfltk and your app, and report whether the error disappears?

```
diff --git a/src/drivers/Wayland/Fl_Wayland_Window_Driver.cxx b/src/drivers/Wayland/Fl_Wayland_Window_Driver.cxx
index 52080c705..8e9f9392b 100644
--- a/src/drivers/Wayland/Fl_Wayland_Window_Driver.cxx
+++ b/src/drivers/Wayland/Fl_Wayland_Window_Driver.cxx
@@ -450,6 +450,7 @@ void Fl_Wayland_Window_Driver::hide() {
       wld_win->subsurface = NULL;
     }
     if (wld_win->kind == DECORATED) {
+libdecor_frame_set_visibility(wld_win->frame, false);
       libdecor_frame_unref(wld_win->frame);
       wld_win->frame = NULL;
       wld_win->xdg_surface = NULL;
```

**Comment by nohsheef** [comment]
@ManoloFLTK will do, might take a moment

**Comment by nohsheef** [comment]
Running `test/pack` in unpatched 4dffa31520f66fe0baea574ea391ec58a77eaa49 and closing window with sway master, gives:

```
Fatal error while communicating with the Wayland server: Protocol error
```

Running `test/pack` after patching `Fl_Wayland_Window_Driver::hide()` and closing window with sway master, gives:

```
Segmentation fault (core dumped)
```

I haven't patched fltk-rs with your fix, will do later today since it requires build fiddling.

**Comment by nohsheef** [comment]
Running `test/pack` in unpatched 4dffa31520f66fe0baea574ea391ec58a77eaa49 and closing window with sway master, gives:

```
Fatal error while communicating with the Wayland server: Protocol error
```

Running `test/pack` after patching `Fl_Wayland_Window_Driver::hide()` and closing window with sway master, gives:

```
Fatal error while communicating with the Wayland server: Protocol error
```

I haven't patched fltk-rs with your fix, will do later today since it requires build fiddling.

**Comment by ManoloFLTK** [comment]
OK. That was an attempt to send a call to `zxdg_toplevel_decoration_v1::destroy` as suggested above.

Thus, to reproduce the error, I have to build sway from source taken from its git repo, master branch. Is that correct?

**Comment by nohsheef** [comment]
Correct, 1.8.1 is silent and sway git complains on window kill.

Is `Wayland server: Protocol error` a sufficient repro, or do I have to figure out how fltk-rs seems to additionally trigger `zxdg_toplevel_decoration_v1@32: error 2:`?

**Comment by nohsheef** [comment]
I don't know if the behavior depends on wlroots master as well. When I build sway from git, I also use a local, statically linked wlroots from latest git. If you have trouble with that, I can help. Or you can use one of the sway-git build recipes (e.g. Arch AUR).

**Comment by ManoloFLTK** [comment]
@nohsheef Could I ask you to make another test with your sway master (ahead of Linux packages)? These are the commands to perform:

```
git clone https://gitlab.freedesktop.org/libdecor/libdecor
cd libdecor
git reset --hard bb2ec856
meson build
meson compile -C build
cd build
export LIBDECOR_PLUGIN_DIR=$PWD/src/plugins/gtk
./demo/libdecor-demo &
```

and type the window-closing shortcut in the demo window which appeared in Sway.
Please, report here if the error appears. If it does, then it shows the error is in `libdecor`, a piece of software bundled by FLTK to handle window decoration.

**Comment by nohsheef** [comment]
- verified that fltk-rs hello world triggers full error incl `zxdg_toplevel_decoration_v1@32`
- verified that fltk/test/pack triggers only `Fatal error while communicating with the Wayland server`
- executed your libdecor steps and saw __no errors__

**Comment by ManoloFLTK** [comment]
@nohsheef Thanks for this test with libdecor. Unfortunately, I forgot the libdecor-demo program does not really perform the full window closing dialog between the client program and the Sway compositor because it calls exit(0) immediately. I would therefore like to ask you to perform a last test. Please, apply this patch to the libdecor source code, and then re-build and re-run the libdecor-demo app. Report here what happens when you type the window-closing shortcut. TIA.
```
diff --git a/demo/demo.c b/demo/demo.c
index 26d955c..85dbe23 100644
--- a/demo/demo.c
+++ b/demo/demo.c
@@ -1150,11 +1150,15 @@ handle_configure(struct libdecor_frame *frame,
 	redraw(window);
 }
 
+static bool was_closed = false;
+
 static void
 handle_close(struct libdecor_frame *frame,
 	     void *user_data)
 {
-	exit(EXIT_SUCCESS);
+	//exit(EXIT_SUCCESS);
+	libdecor_frame_unref(frame);
+	was_closed = true;
 }
 
 static void
@@ -1310,7 +1314,8 @@ main(int argc,
 	gettimeofday(&start, NULL);
 	libdecor_frame_set_min_content_size(window->frame, 15 * chk, 10 * chk);
 
-	while (libdecor_dispatch(context, timeout) >= 0) {
+	while (libdecor_dispatch(context, timeout) > 0) {
+		if (was_closed) timeout = 0;
 		if (timed_quit) {
 			gettimeofday(&now, NULL);
 			if (now.tv_sec >= start.tv_sec + 2) {
```

**Comment by nohsheef** [comment]
- verified I'm in the right sway version and fltk-rs hello world provokes both errors
- no protocol error after rebuilding with the diff applied and pressing sway's kill binding
- pressing escape in libdecor demo exits too and prints _close_ to stdout

**Comment by ManoloFLTK** [comment]
@nohsheef  OK. Thanks for this test. That gives me a lead to follow: make sure FLTK closes libdecor-decorated windows exactly how pure libdecor does it.

**Comment by jbeich** [comment]
> Thus, to reproduce the error, I have to build sway from source taken from its git repo, master branch. Is that correct?

FWIW, FreeBSD has the latest Sway from Git `master` packaged as [sway-devel](https://www.freshports.org/x11-wm/sway-devel).

**Comment by ManoloFLTK** [comment]
@jbeich Many thanks for this information. Having this advanced version of Sway would allow me to test the changes in FLTK I've prepared. 

But where can I find package sway-devel?
Command
`sudo pkg install sway-devel`
produces
`No packages available to install matching 'sway-devel' have been found in the repositories`

**Comment by ManoloFLTK** [comment]
@nohsheef Could you, please, apply the attached patch to the source code of FLTK, rebuild FLTK, and try test/pack with Sway's kill binding. TIA

[issue750.patch](https://github.com/fltk/fltk/files/12088753/issue750.patch)

**Comment by jbeich** [comment]
> But where can I find package sway-devel?

Depends on release/architectures tuple and [package branch](https://docs.freebsd.org/en/books/handbook/ports/#quarterly-latest-branch). FreshPorts has a [nice UI](https://www.freshports.org/x11-wm/sway-devel#packages) to show current status. If a package is missing it may be due to intermittent build failures, the cluster machine being slow or shared to build for another release/architecture.

All builds are automatic and build as a set to avoid ABI mismatch. Builds are scheduled to start every Tue/Thu/Sat/Sun at 01:00 UTC but if the previous build hasn't finished the next one isn't started. For example, `FreeBSD:13:aarch64` on `/quarterly` still lacks `sway-devel` because when `2023Q3` branched on 2023-07-01 the machine was [busy](https://pkg-status.freebsd.org/ampere1/build.html?mastername=124arm64-quarterly&build=bee14067723b) building `2023Q2` for `FreeBSD:12:aarch64`, only [started](https://pkg-status.freebsd.org/ampere1/build.html?mastername=131arm64-quarterly&build=8898bf465b47) with `2023Q3` on 2023-07-15 and hasn't finished yet. Looking at [pkg-status](https://pkg-status.freebsd.org/?all=1&type=package) aarch64 has only 3 machines.

**Comment by nohsheef** [comment]
@ManoloFLTK unless I did something wrong, the protocol error persists. I did a clean rebuild and retested a 2nd time, and here's the branch with your patch applied which I've built from: https://github.com/nohsheef/fltk/compare/master...issue750

**Comment by ManoloFLTK** [comment]
@nohsheef  Thanks for testing. I've now installed Sway-master under FreeBSD and can reproduce the error target of this issue with FLTK. I'll try to fix it and will post code on https://github.com/ManoloFLTK/fltk if I need to ask you to perform other tests.

**Comment by ManoloFLTK** [comment]
@nohsheef Being able to reproduce the issue, I have studied it in detail. My conclusion is that it's not a bug in FLTK, but a bug :
1. either in Sway-master that incorrectly deals with unstable protocol ["xdg decoration"](https://wayland.app/protocols/xdg-decoration-unstable-v1) which "allows a compositor to announce support for server-side decorations";
2. or in libdecor which does something bad specifically with Sway-master in its interactions with "xdg decoration".

I believe case 1 is more likely than case 2.

The proof that the bug is not in FLTK comes from the fact that the error can be produced with libdecor-only code (modifying libdecor's demo.c), that is, with absolutely zero use of FLTK.

The observation is
- no problem when the modified demo runs on Sway from the standard FreeBSD package;
- error `error 2: xdg_toplevel destroyed before xdg_toplevel_decoration`
when the same demo runs on Sway-master.

The modified demo program can be built with
```
git clone https://gitlab.freedesktop.org/ManoloFLTK/libdecor libdecor-fltk
cd libdecor-fltk
git checkout issue750
meson build
meson compile -C build
cd build
export LIBDECOR_PLUGIN_DIR=$PWD/src/plugins/gtk
./demo/libdecor-demo &
```

This program opens 2 identical windows, instead of one, so that it continues running after one window gets closed. It runs without problems with standard Sway, and stops with error with Sway-master as soon as one window is closed. Closing can be done by the Sway kill command or by hitting escape. Both produce the same error.

@jbeich : would you interpret these observations differently for the location of the bug ?

**Comment by nohsheef** [comment]
@ManoloFLTK thanks for the investigation. Since I'm running sway and don't rely on CSD, would it be possible to disable libdecor at build time?

**Comment by jbeich** [comment]
> @jbeich : would you interpret these observations differently for the location of the bug ?

I'm not familar with the API, so don't know if your example is correct. Other libdecor consumers appear to use it only for CSD, not for hybrid CSD + SSD or SSD via libdecor instead of xdg-decoration directly.

**Comment by ManoloFLTK** [comment]
FLTK can produce borderless windows. These don't use libdecor at all. They are obtained calling
`window->border(0)`.
So, it's possible to use FLTK without libdecor being ever used if only borderless windows are created. But these are just rectangles on the screen. They can't be resized. They don't tile the screen. They don't reply to the Sway's kill command.

**Comment by ManoloFLTK** [comment]
> Other libdecor consumers appear to use it only for CSD, not for hybrid CSD + SSD or SSD via libdecor instead of xdg-decoration directly.

Do you mean that other libdecor consumers either always work in CSD mode, even with Sway and KWin, or don't use libdecor to negotiate with the compositor SSD mode using the xdg-decoration protocol?

Unless I'm wrong, I believe FLTK uses libdecor the way it was intended to be used, that is, letting libdecor decide whether to use CSD or SSD and using the xdg decoration protocol for that decision. It's important to note that libdecor's public API completely hides the mechanism that produces window decorations. Consequently, FLTK doesn't know if it runs in CSD or SSD mode: that's not accessible via the public libdecor API.

**Comment by jbeich** [comment]
> don't use libdecor to negotiate with the compositor SSD mode using the xdg-decoration protocol?

Better ask upstream libdecor. I mainly work on packaging.

In FreeBSD libdecor currently only has 3 consumers:
- `sdl2` added libdecor after xdg-decoration, so libdecor is unused for non-CSD without [`SDL_VIDEO_WAYLAND_PREFER_LIBDECOR`](https://github.com/libsdl-org/SDL/commit/9c2f46b0d5a5).
- `xwayland-devel` when rootful with `-decorate` for CSD, otherwise SSD because it doesn't use xdg-decoration thus cannot opt-out
- `blender` [limits libdecor to GNOME](https://github.com/blender/blender/commit/f0e1089a33f4) but in practice `gtk_shell` is supported by Mutter, Phoc (wlroots), Wayfire (wlroots).

**Comment by jbeich** [comment]
Oops, Sway appears to use SSD by default without xdg-decoration while Wayfire shows borderless windows. I should probably keep silent instead of promoting confusion.

**Comment by ManoloFLTK** [comment]
> Le 20 juil. 2023 à 18:35, Jan Beich ***@***.***> a écrit :
> 
> 
> Oops, Sway appears to use SSD by default without xdg-decoration while Wayfire shows borderless windows. I should probably keep silent instead of promoting confusion.
> —

Many thanks Jan for having pointed me to the sway-devel package which allowed me to reproduce the problem.

This is an image of what I get when I create a Sway window without help from libdecor and thus without use of "xdg decoration":




No decoration at all, neither from client nor from Sway.

Best
Manolo

**Comment by ManoloFLTK** [comment]
@nohsheef Given the observations made above, I believe nothing can be done within FLTK to fix the problem reported here. Unless you have other questions, I would suggest you close this issue. You may want to report the error you discovered to Sway developpers.

**Comment by nohsheef** [comment]
Sure, I can close it. I'm a little hesitant about reporting it to sway developers given @Nefsen402's evaluation.

**Comment by ManoloFLTK** [comment]
@nohsheef You're right. I believe now it might be a bug in libdecor, at the app level as wrote @Nefsen402 . I might report that to libdecor.

**Comment by erco77** [comment]
Just observing from afar: if "kill" from sway's hotkey really means "kill", as in sending a trappable signal(2) to the application with kill(2), then perhaps the error message can be avoided by trapping the signal in the application at the top of `main()`, with something like `signal(sig_num, some_func);` and doing a normal exit(0)? Where `sig_num` might be  SIGINT, SIGQUIT, SIGTERM, or some such.

fluid does this with SIGINT, so perhaps an easy check would be: 1) run fluid, 2) use sway's hotkey to kill it, 3) check if the error is printed.

It's possible sway is using something other than SIGINT, in which case add traps for other common termination signals, e.g. in `fluid.cxx` add these lines:
```
 #ifndef _WIN32
   signal(SIGINT,sigint);
+  signal(SIGTERM,sigint);
+  signal(SIGQUIT,sigint);
+  signal(SIGHUP,sigint);
 #endif

```

**Comment by ManoloFLTK** [comment]
@erco77 It's not a kill, as visible with a 2-window situation where it closes the active window and lets the program continue happily. It's just Sway's replacement for the window closing button that does not exist as a button in the titlebar.
Anyway, I've found the fix which should go to libdecor and have filed a libdecor merge request for it.

**Comment by nohsheef** [comment]
@ManoloFLTK applying your [libdecor patch](https://github.com/fltk/fltk/commit/46e6dbac528f9551ae68598306db3eb0fa84310b) to my copy of fltk seems to fix the protocol error.

**Comment by ManoloFLTK** [comment]
@nohsheef Yes, the 46e6dba commit fixes the issue reported here. That's its purpose. But this is a commit to [libdecor](https://gitlab.freedesktop.org/libdecor/libdecor/). The policy regarding libdecor at FLTK is to bundle **unchanged** libdecor source code inside the FLTK source tree. Therefore, this commit won't be made to FLTK. We just have to wait until it's committed inside the libdecor git repo, and then it will be inherited by FLTK. At least 6 FLTK-advocated MR's have in the past been accepted and committed to libdecor, and subsequently inherited by FLTK. I have good hope [libdecor MR # 128](https://gitlab.freedesktop.org/libdecor/libdecor/-/merge_requests/128) will make it too, unless it's an incorrect solution, or it's found the error lies in sway-master. Please, reconsider whether this issue needs been re-opened.

**Comment by nohsheef** [comment]
Closing as resolved.

**Comment by nohsheef** [comment]
@ManoloFLTK to be precise, once libdecor lands your fix, will you sync libdecor completely and as is with the state of upstream? Or do you pick diffs?

**Comment by nohsheef** [comment]
I'm asking about the libdecor vendoring process because I've been debugging `libdecor_get_cursor_settings()` dbus code path and how to make it work in my environment. To that end, I've adapted the code into a stand-alone test. The short environment variable code path works as expected, but dbus doesn't yet, although cursor theme and size are configured in the right gnome schema and are picked up by gtk apps.  So I've been wondering if the dbus code path could use a fallback call to the non-dbus code path on failure. This is off-topic and I just wanted to explain why I asked about the libdecor vendoring process. That's all.

**Comment by Albrecht-S** [comment]
@nohsheef
```
$ git log -9 --author=ManoloFLTK --pretty='%h %ci %an - %s' -- libdecor/
3166416e4 2023-06-13 07:42:44 +0200 ManoloFLTK - libdecor: update with upstream source code as of 2023-jun-6
87dec3af9 2023-06-07 15:58:12 +0200 ManoloFLTK - libdecor: update with upstream source code as of 2023-may-26
cad6e6329 2023-05-02 11:35:28 +0200 ManoloFLTK - libdecor: update with upstream source code as of 2023-may-2
cdd1566cf 2023-04-30 18:41:49 +0200 ManoloFLTK - Allow building Wayland platform without dbus (#726)
64ac7b47b 2023-03-29 15:25:29 +0200 ManoloFLTK - Improve detection of the location of Wayland protocol files across OSes
02adfc056 2023-02-16 08:56:37 +0100 ManoloFLTK - Wayland: remove libdecor/demo/* from FLTK source tree
83d9a2c63 2022-12-29 15:26:28 +0100 ManoloFLTK - libdecor: update from upstream (3f3e5e1d) as of 29-dec-2022
4d8ae8ccd 2022-12-08 12:41:41 +0100 ManoloFLTK - Libdecor: update to upstream commit 25f29fef 3 dec 2022
a5d2b5ed4 2022-11-03 08:16:49 +0100 ManoloFLTK - libdecor: update from source git repo (commit e87dcfda)
```

In the "early days" of our Wayland port we had some own mods in the libdecor sources, but - as Manolo wrote - the current policy is to bundle the libdecor code as is.

**Comment by ManoloFLTK** [comment]
@nohsheef FLTK copies several files from the libdecor source tree to its own source tree **without any change to these files**. But not all the libdecor source tree is copied. For example, meson-related files are not copied. File `cursor-settings.c` which contains the code of function `libdecor_get_cursor_settings()` is copied, unchanged. This file provides 2 versions of this function to be used with or without DBUS. At FLTK build time, it's determined whether DBUS is present on the build system, and this determines which of the 2 versions of `libdecor_get_cursor_settings()` will FLTK use. I try to follow frequently libdecor updates and to transfer them to FLTK.

**Comment by nohsheef** [comment]
@ManoloFLTK I got to the bottom of the dbus internal error on my desktop and so far fail to provide the relevant endpoints for the dbus message to be processed correctly. Since the environment variable code path just works without the complexity of dbus, are you open to a build flag to make dbus opt-out? That way, I would also remove a dependency which isn't crucial for libdecor's core functionality.

**Comment by ManoloFLTK** [comment]
@nohsheef DBUS is not a requirement because building without it is possible. To build FLTK without DBUS use even if DBUS is present on the building system can be achieved commenting out line # 514 of file `src/CMakeLists.txt` of the FLTK source tree :
```
  pkg_check_modules(DBUS dbus-1)
```
I'm a bit wary to add another build option since there are already many.

**Comment by nohsheef** [comment]
I can relate to that. The fallback proposed in https://gitlab.freedesktop.org/libdecor/libdecor/-/issues/54 would be a good idea for resilience against dbus issues anywhere outside GNOME and solve the same issue (minus runtime dependency on dbus).

**Comment by ManoloFLTK** [comment]
@nohsheef Libdecor also offers a way to enforce use of client-side decoration (CSD) when the Wayland compositor, as Sway does, would use server-side decoration (SSD). That's obtained setting an environment variable :
 ```
export LIBDECOR_FORCE_CSD=1
```
After that, the issue reported here disappears because it's SSD-specific. It's possible to close the active window with Sway's kill shortcut. No Wayland error occurs when using Sway master.

Libdecor can draw window titlebars either GTK styled or using Cairo. The selection between these 2 possibilities is done at FLTK build time: if package libgtk-3-dev is available, libdecor will use GTK, otherwise it will use Cairo to draw window titlebars. Moreover, it's possible to ask FLTK to use Cairo for titlebars even in the presence of libgtk-3-dev with CMake option OPTION_ALLOW_GTK_PLUGIN=0

**Comment by ManoloFLTK** [comment]
@nohsheef MR !128 to libdecor has beeen accepted and merged into the libdecor source repo. This has subsequently been merged into the FLTK repo (afc60b2). With this, you can now use FLTK with sway-devel without crash.
